File: frontend/types/comment.ts
import type { Auction } from "@/types/auction";
import type { ChatMessage } from "@/types/chat";

export interface AuctionComment {
  id: string;
  auction_id: string;
  user_id: string;
  author_name: string;
  body: string;
  created_at: string;
}

// Everything the auction-room WebSocket can send.
export type RoomMessage =
  | { type: "snapshot" | "bid" | "updated" | "ended"; auction: Auction }
  | { type: "comment"; comment: AuctionComment }
  | { type: "chat"; message: ChatMessage };

File: backend/app/models/comment.py
import uuid
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, String, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.models.user import User


class AuctionComment(Base):
    """One live-chat message in an auction room. Posted by any user who can view the auction."""

    __tablename__ = "auction_comments"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    auction_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("auctions.id", ondelete="CASCADE"), index=True
    )
    user_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"))
    body: Mapped[str] = mapped_column(String(1000))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

    # Carries the author's name inline so a message never shows a bare UUID.
    user: Mapped[User] = relationship(lazy="selectin")
